racker/salus-telemetry-label-management
# Create node presence keys in etcd

# Resolves

https://jira.rax.io/browse/SALUS-73

# What
(Stolen from Jira... will update later)
The Ambassador receives the node id from the envoy on attachment and uses it to create/update the etcd expected/active/identifier keys as described here:

https://one.rackspace.com/display/monitoring/Salus+Node+Presence+Monitoring#SalusNodePresenceMonitoring-EtcdKeysets

It also needs to update the key's lease with each heartbeat, and revoke the lease when the heartbeat goes away.


# How

The `tenant:identifier:identifierValue` string gets hashed and is used as part of the key in both `/nodes/expected/hash` and `/nodes/active/hash` when an envoy connects.  The value stored is a `NodeInfo` object - https://github.com/racker/salus-telemetry-model/pull/3

The `expected` key is created without a lease so it will live forever.

The `active` key is created using the same lease that the `envoyLabelManagement.registerAndSpreadEnvoy` method creates.  This means the same keepAlive will take care of ensure the `active` key remains connected.  It also keeps the two different "register' keys in sync since any keys linked to the lease will all be removed as soon as the lease expires.

## How to test

Run the `testRegister` test.

or

Run the Ambassador locally, run an envoy locally, look in etcd and see the keys there.  Stop the envoy and see the `active` key disappear.

# Why

To do...

# TODO

- `AmbassadorHashing` should move to `salus-common` and probably needs to be tweaked.
